File: provision/go/internal/environment/scripted_input.go
package environment

type ScriptedInput struct {
	queued []string
}

func NewScriptedInput(lines []string) *ScriptedInput {
	queued := make([]string, len(lines))
	copy(queued, lines)

	return &ScriptedInput{queued: queued}
}

func (input *ScriptedInput) ReadKey() (string, bool, error) {
	if len(input.queued) == 0 {
		return "", false, nil
	}

	value := input.queued[0]
	input.queued = input.queued[1:]

	return value, true, nil
}

func (input *ScriptedInput) ReadLine(_ string) (string, error) {
	if len(input.queued) == 0 {
		return "", nil
	}

	value := input.queued[0]
	input.queued = input.queued[1:]

	return value, nil
}

File: pkgs-ts/console/src/environment/scripted-input.ts
import { parseScriptedInputLines } from '#console/environment/validators/scripted-input';
import type { PromptInput } from '#console/types';

export const createScriptedInput = (lines: string[]): PromptInput => {
	const queued = parseScriptedInputLines(lines);

	return {
		async readKey(): Promise<string | null> {
			return queued.shift() ?? null;
		},
		async readLine(): Promise<string> {
			return queued.shift() ?? '';
		},
	};
};
